### ISSUE TITLE
hasattr(obj, '__slots__') does not allow access to obj.__slots__

### ISSUE THREAD

**Issue by effigies** [microsoft/pyright#7363]
**Describe the bug**
`if hasattr(obj, '__slots__')` is not sufficient for pyright to infer the existence of `obj.__slots__` for `obj` with type `object`.

**Code or Screenshots**
```python
from typing import Any
import attrs

def todict(obj: object) -> dict[str, Any]:
    if attrs.has(type(obj)):
        return attrs.asdict(obj)
    if hasattr(obj, '__slots__'):
        return {attr: getattr(obj, attr) for attr in obj.__slots__}
    return obj.__dict__
```

```console
/home/chris/Sandbox/typing/slots.py
  /home/chris/Sandbox/typing/slots.py:8:58 - error: Cannot access member "__slots__" for type "object"
    Member "__slots__" is unknown (reportAttributeAccessIssue)
1 error, 0 warnings, 0 informations 
```

I verified that adding `isinstance(obj.__slots__, Sequence)` doesn't help, so it seems to be a failure to accept `__slots__` as existing, not uncertainty about what type `__slots__` is.

**VS Code extension or command-line**

CLI: `pyright 1.1.351`

**Comment by erictraut** [comment]
Pyright is behaving as intended here, so I don't consider this a bug.

The `object` class does not have an attribute named `__slots__` according to the typeshed class definition for `object`. The check `hasattr(obj, '__slots__')` is not a supported [type guard pattern](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=type-guards), but even if it was, the type of `obj` cannot be narrowed further than `object`. 

To make this work with static typing, you can define your own protocol class like the following:

```python
from typing import Any, Iterable, Protocol
import attrs


class SupportsSlots(Protocol):
    __slots__: None | Iterable[str]

def todict(obj: SupportsSlots) -> dict[str, Any]:
    if attrs.has(type(obj)):
        return attrs.asdict(obj)
    if hasattr(obj, "__slots__") and obj.__slots__ is not None:
        return {attr: getattr(obj, attr) for attr in obj.__slots__}
    return obj.__dict__
```

Note: I also recommend checking for `None` because that's a legal value for `__slots__`.
